src/OrchardCore.Modules/Orchard.Lucene/LuceneIndexManager.cs

```diff
@@ -0,0 +1,347 @@
+using Directory = System.IO.Directory;
+using LuceneNetCodecs = Lucene.Net.Codecs;
+
+namespace Orchard.Lucene
+{
+    /// <summary>
+    /// Provides methods to manage physical Lucene indices.
+    /// This class is provided as a singleton to that the index searcher can be reused across requests.
+    /// </summary>
+    public class LuceneIndexManager : IDisposable
+    {
+        private readonly IHostingEnvironment _hostingEnvironment;
+        private readonly string _rootPath;
+        private readonly DirectoryInfo _rootDirectory;
+
+        private ConcurrentDictionary<string, IndexReaderPool> _indexPools = new ConcurrentDictionary<string, IndexReaderPool>(StringComparer.OrdinalIgnoreCase);
+        private ConcurrentDictionary<string, IndexWriter> _writers = new ConcurrentDictionary<string, IndexWriter>(StringComparer.OrdinalIgnoreCase);
+
+        private static LuceneVersion LuceneVersion = LuceneVersion.LUCENE_48;
+
+        static LuceneIndexManager()
+        {
+            SPIClassIterator<LuceneNetCodecs.Codec>.Types.Add(typeof(LuceneNetCodecs.Lucene46.Lucene46Codec));
+            SPIClassIterator<LuceneNetCodecs.PostingsFormat>.Types.Add(typeof(LuceneNetCodecs.Lucene41.Lucene41PostingsFormat));
+            SPIClassIterator<LuceneNetCodecs.DocValuesFormat>.Types.Add(typeof(LuceneNetCodecs.Lucene45.Lucene45DocValuesFormat));
+        }
+
+        public LuceneIndexManager(
+            IHostingEnvironment hostingEnvironment,
+            IOptions<ShellOptions> shellOptions,
+            ShellSettings shellSettings
+            )
+        {
+            _hostingEnvironment = hostingEnvironment;
+
+            _rootPath = Path.Combine(
+                _hostingEnvironment.ContentRootPath,
+                shellOptions.Value.ShellsRootContainerName,
+                shellOptions.Value.ShellsContainerName,
+                shellSettings.Name, "Lucene");
+
+            _rootDirectory = Directory.CreateDirectory(_rootPath);
+        }
+
+        public void CreateIndex(string indexName)
+        {
+            var path = new DirectoryInfo(Path.Combine(_rootPath, indexName));
+
+            if (!path.Exists)
+            {
+                path.Create();
+            }
+
+            Write(indexName, _ => { }, true);
+        }
+
+        public void DeleteDocuments(string indexName, IEnumerable<string> contentItemIds)
+        {
+            Write(indexName, writer =>
+            {
+                writer.DeleteDocuments(contentItemIds.Select(x => new Term("ContentItemId", x)).ToArray());
+
+                writer.Commit();
+
+                if (_indexPools.TryRemove(indexName, out var pool))
+                {
+                    pool.MakeDirty();
+                }
+            });
+        }
+
+        public void DeleteIndex(string indexName)
+        {
+            lock (this)
+            {
+                if (_writers.TryRemove(indexName, out var writer))
+                {
+                    writer.Dispose();
```

Comment:
An index is only managed by a single tenant on a single box. Unless in the case Azure App Services which shares the filesystem and in this case only the writer will be an issue. So yes it is an issue in this case, but just for the writer. In this case I need to ensure only one tenant per index can write. I will just ensure that writes can fail if there is a write.lock file.